message_ix/tests/test_feature_price_emission.py

```diff
@@ -30,18 +31,75 @@ def model_setup(scen, years, simple_tecs=True):
 def add_two_tecs(scen, years):
     """add two technologies to the scenario"""
     scen.add_set("technology", ["dirty_tec", "clean_tec"])
-    output_specs = ["node", "comm", "level", "year", "year"]
 
     for y in years:
         # the dirty technology is free (no costs) but has emissions
-        tec_specs = ["node", "dirty_tec", y, y, "mode"]
-        scen.add_par("output", tec_specs + output_specs, 1, "GWa")
-        scen.add_par("emission_factor", tec_specs + ["co2"], 1, "tCO2")
+        scen.add_par(
+            "output",
+            make_df(
+                "output",
+                node_loc="node",
+                node_dest="node",
+                technology="dirty_tec",
+                year_vtg=y,
```

Comment:
If using `make_df()` here, then it can accept a sequence (e.g. list) for any of its arguments/the parameter dimensions. So one could remove `for y in years:` and instead do:

```python
common = dict(
   mode="mode",
   technology="dirty_tec",
   value=1.0,
   year_vtg=years,
   year_act=years,
   ...  # All other labels that are the same in `output` and `emission_factor`
)
name = "output"
scen.add_par(
    name,
    make_df(
       name,
       unit="GWa",
       ...  # All other labels that are specific to `output`
       **common
    )
name = "emission_factor"
scen.add_par(...)
```